### PR TITLE
[ENG-6119] api changes for institution-users updates

### PR DESCRIPTION
**Pull request by aaxelb**
## Purpose
update api to serve institutional-user metrics based on the monthly report added in #10722 

## Changes
reusable things added:
- `toggle_view_by_flag` util to ease feature-flagged views
    - also allow non-class views to exist (minor updates to api tests and bases)
- base view `ElasticsearchListView`
    - use `elasticsearch_dsl.Search` as queryset-analogue, for code reuse
    - allows filtering (with 'eq' and 'ne' operators) following `filterable_fields` on the serializer (using `api.base.filters.FilterMixin`)
    - allows sorting, following `default_ordering` and `ordering_fields` on the view
    - allows pagination (with `page` and `page[size]` params), using default paginator
- `api.base.serializers.YearmonthField`
- `MonthlyReport.most_recent_yearmonth` classmethod

specific to this project:
- add feature flag `INSTITUTIONAL_DASHBOARD_2024`
- rename prior institution-user view and serializer with `Old` prefix
- use `toggle_view_by_flag` to toggle new/old views (with new/old serializers)
- add "new" institution-user metrics view (using `ElasticsearchListView`)
- add "new" institution-user metrics serializer

incidental:
- fix: old institution-user metrics tests
    - set INSTITUTIONAL_DASHBOARD_2024 flag inactive for existing institution-user metrics tests
    - remove `time.sleep` calls; use index refreshes instead (saves ~70 sec)
    - unskip tests -- "non-deterministic fails" may be fixed by refreshes?
- avoid clobbering all indexes by running tests
    - rename the `es` pytest marker to `es_metrics` (for clarity)
    - update the effect of that marker:
        - patch a prefix to each metric class's index and template names
        - instead of deleting ALL indexes and index templates, delete only
          those with the patched prefix
    - update `api_tests.metrics.test_raw_metrics` to clean up after itself
      and stop depending on the clobbering

## QA Notes

Please make verification statements inspired by your code and what your code touches.
- Verify
- Verify

What are the areas of risk?

Any concerns/considerations/questions that development raised?

## Ticket
[ENG-6119]

[ENG-6119]: https://openscience.atlassian.net/browse/ENG-6119?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: api/base/elasticsearch_dsl_views.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+from rest_framework import generics
+from rest_framework import exceptions as drf_exceptions
+from rest_framework.settings import api_settings as drf_settings
+
+from api.base.filters import FilterMixin
+from api.base.views import JSONAPIBaseView
+
+
+class ElasticsearchListView(FilterMixin, JSONAPIBaseView, generics.ListAPIView, abc.ABC):
+    '''abstract view class using `elasticsearch_dsl.Search` as a queryset-analogue
```
Comment (Johnetordoff) [inline api/base/elasticsearch_dsl_views.py:14]: Might want to point out when to use this, how this improves on earlier ES6 views.

File: api/base/elasticsearch_dsl_views.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        return (_elastic_sort_field if _ascending else f'-{_elastic_sort_field}')
+
+    def __add_search_filter(
+        self,
+        search: edsl.Search,
+        elastic_field_name: str,
+        operator: str,
+        value: str,
+    ) -> edsl.Search:
+        match operator:  # operators from FilterMixin
```
Comment (Johnetordoff) [inline api/base/elasticsearch_dsl_views.py:88]: Are we sure there's no [lt]/[lte]/[gt]/[gte] ? Thinking about if admins want custom csv files in certain date ranges they may want to filter by date. I'm not sure it's necessary from this view, not for the user's tab anyway.

File: api/base/serializers.py [hunk 3]
```diff
@@ -927,8 +928,8 @@ def to_representation(self, value):
         relationship = format_relationship_links(related_url, self_url, related_meta, self_meta)
         if related_url:
             resolved_url = resolve(related_path)
-            related_class = resolved_url.func.view_class
-            if issubclass(related_class, RetrieveModelMixin):
+            related_class = getattr(resolved_url.func, 'view_class', None)
```
Comment (Johnetordoff) [inline api/base/serializers.py:931]: Not sure why this change was made, but I know there was a good reason. Maybe add a comment or assert something to demonstrate this conditional pertains to the new views/features.

File: api/institutions/serializers.py [hunk 4]
```diff
@@ -306,3 +307,43 @@ def get_absolute_url(self, obj):
                 'version': 'v2',
             },
         )
+
+
+class NewInstitutionUserMetricsSerializer(JSONAPISerializer):
```
Comment (Johnetordoff) [inline api/institutions/serializers.py:312]: Obviously calling something `New`/`Old` is kind of silly considering it's not versioned in anyway, but hey go for it. At least it's clear what we are doing with this naming convention, it can always be renamed to `SecondNewest` when it's no longer new in 2034. :)

File: api/institutions/serializers.py [hunk 5]
```diff
@@ -270,7 +271,7 @@ def get_absolute_url(self, obj):
         )
 
 
-class InstitutionUserMetricsSerializer(JSONAPISerializer):
+class OldInstitutionUserMetricsSerializer(JSONAPISerializer):
```
Comment (Johnetordoff) [inline api/institutions/serializers.py:274]: Maybe add docstring explain `New`/`Old` naming convention, dashboard flag names and purpose.

File: api_tests/base/test_views.py [hunk 6]
```diff
@@ -43,8 +43,9 @@
         if hasattr(patt, 'url_patterns'):
             # Namespaced list of patterns
             for subpatt in patt.url_patterns:
-                VIEW_CLASSES.append(subpatt.callback.cls)
-        else:
+                if hasattr(subpatt.callback, 'cls'):
```
Comment (Johnetordoff) [inline api_tests/base/test_views.py:46]: This is another line I can't quite tell why this is needed, but I'm sure there's a good reason for it. Maybe use a named function or throw a comment in there.

File: api/base/serializers.py [hunk 7]
```diff
@@ -959,7 +960,7 @@ def to_representation(self, value):
                         related_id = resolved_url.kwargs['institution_id']
                         related_type = 'institution-summary-metrics'
                     elif related_type == 'collections' and related_class.view_name == 'collection-submission-detail':
-                        related_id = f'{resolved_url.kwargs['collection_submission_id']}-{resolved_url.kwargs['collection_id']}'
+                        related_id = f"{resolved_url.kwargs['collection_submission_id']}-{resolved_url.kwargs['collection_id']}"
```
Comment (Johnetordoff) [inline api/base/serializers.py:963]: I think this should be single quotes? Confused.

File: api/institutions/views.py [hunk 8]
```diff
@@ -521,3 +528,52 @@ def get_default_queryset(self):
...
+        if _yearmonth is None:
+            return None
+        return (
+            InstitutionalUserReport.search()
+            .filter('term', report_yearmonth=str(_yearmonth))
+            .filter('term', institution_id=self.get_institution()._id)
+        )
+
+
+institution_user_metrics_list_view = view_toggled_by_feature_flag(
```
Comment (Johnetordoff) [inline api/institutions/views.py:574]: @futa-ikeda Do we have a clean-up ticket for after the flag is flipped? This is fine for feature flipping, but is a weird import side-effect in the long term.

**Review by Johnetordoff — COMMENTED** [review]
I don't really have any hard changes, just a couple questions and ideas. Good work! 💯

File: api/base/elasticsearch_dsl_views.py (hunk 2 above)
Comment (aaxelb) [reply api/base/elasticsearch_dsl_views.py:88]: i'm open to adding range filters if/when we have a use for it (would be nice to make this base view a more generally reusable thing), but there's been no ask for it yet and i'm trying to embrace at least the good parts of agile :)

(if we decide to offer an institution-summary history and/or use this to improve the unstable/terrible `/_/metrics/reports/` api, then yeah, range queries for sure)

File: api/institutions/serializers.py (hunk 4 above)
Comment (aaxelb) [reply api/institutions/serializers.py:312]: really hoping we manage to get around to deleting code at some point...

File: api_tests/base/test_views.py (hunk 6 above)
Comment (aaxelb) [reply api_tests/base/test_views.py:46]: another "allow view functions to exist" -- will comment

File: api/institutions/views.py (hunk 8 above)
Comment (aaxelb) [reply api/institutions/views.py]: added a cleanup ticket: https://openscience.atlassian.net/browse/ENG-6223
